### PR TITLE
SONARJAVA-1461 Change of method contract nullability

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      if (nonNullVsNull(paramSymbol, overrideeParamSymbol) || nonNullVsNull(overrideeParamSymbol, paramSymbol)) {
+        reportIssue(methodTree.parameters().get(i), "The \"" + paramSymbol.name() + "\" parameter nullability is different in the superclass method, and that should not be changed.");
+      }
+    }
+    if (nonNullVsNull(methodTree.symbol(), overridee) || nonNullVsNull(overridee, methodTree.symbol())) {
+      reportIssue(methodTree.returnType(), "The return value nullability of this method is different in the superclass, and that should not be changed.");
+    }
+  }
+
+  private boolean nonNullVsNull(Symbol sym1, Symbol sym2) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: MAJOR Make "nonNullVsNull" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  private void checkContractChange(MethodTreeImpl methodTree, JavaSymbol.MethodJavaSymbol overridee) {
+    if (methodTree.isEqualsMethod() && methodTree.parameters().get(0).symbol().metadata().isAnnotatedWith("javax.annotation.Nonnull")) {
+      reportIssue(methodTree.parameters().get(0), "Equals method should accept null parameters and return false.");
+      return;
+    }
+    for (int i = 0; i < methodTree.parameters().size(); i++) {
+      Symbol paramSymbol = methodTree.parameters().get(i).symbol();
+      Symbol overrideeParamSymbol = overridee.getParameters().scopeSymbols().get(i);
+      if (nonNullVsNull(paramSymbol, overrideeParamSymbol) || nonNullVsNull(overrideeParamSymbol, paramSymbol)) {
+        reportIssue(methodTree.parameters().get(i), "The \"" + paramSymbol.name() + "\" parameter nullability is different in the superclass method, and that should not be changed.");
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/resolve/Symbols.java [hunk 3]
```diff
@@ -124,6 +124,12 @@ public String toString() {
     unknownType.interfaces = ImmutableList.of();
     unknownSymbol.type = unknownType;
     unknownMethodSymbol = new JavaSymbol.MethodJavaSymbol(0, "!unknown!", unknownSymbol) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Symbols.java]: MAJOR Reduce this anonymous class number of lines from 22 to at most 20, or make it a named class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1188)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    if (methodSymbol.isMethodSymbol()) {
+      JavaSymbol.MethodJavaSymbol overridee = ((JavaSymbol.MethodJavaSymbol) methodSymbol).overriddenSymbol();
+      if (overridee != null && overridee.isMethodSymbol()) {
+        checkContractChange(methodTree, overridee);
+      }
+    }
+  }
+
+  private void checkContractChange(MethodTreeImpl methodTree, JavaSymbol.MethodJavaSymbol overridee) {
+    if (methodTree.isEqualsMethod() && methodTree.parameters().get(0).symbol().metadata().isAnnotatedWith("javax.annotation.Nonnull")) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: In term of code coverage, the case of having the equals method without changing the contract is not present in the test cases.

File: java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTreeImpl methodTree = (MethodTreeImpl) tree;
+    Symbol.MethodSymbol methodSymbol = methodTree.symbol();
+    if (methodSymbol.isMethodSymbol()) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: As you know it's a method symbol from getting it the ``methodTree``,  why testing it again? Is it to cover the case of unknown symbol? If it's the case, I would prefer to test it by using ``!methodSymbol.isUnknown()``.
Moreover, this case is not covered by your test cases. The same remarks apply to the overridee below.

File: java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    for (int i = 0; i < methodTree.parameters().size(); i++) {
+      Symbol paramSymbol = methodTree.parameters().get(i).symbol();
+      Symbol overrideeParamSymbol = overridee.getParameters().scopeSymbols().get(i);
+      if (nonNullVsNull(paramSymbol, overrideeParamSymbol) || nonNullVsNull(overrideeParamSymbol, paramSymbol)) {
+        reportIssue(methodTree.parameters().get(i), "The \"" + paramSymbol.name()
+            + "\" parameter nullability is different in the superclass method, and that should not be changed.");
+      }
+    }
+    if (nonNullVsNull(methodTree.symbol(), overridee) || nonNullVsNull(overridee, methodTree.symbol())) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: What about the absence of annotation in the current method when overidee is annotated? Should we cover the case? (i.e. Overridee first param being annotated with ``@Nonnull`` when current method is not annotated at all)

Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ChangeMethodContractCheck.java]: I would not raise an issue in taht case for now as long as annotation from package are not handled.

File: java-squid/src/main/java/org/sonar/java/resolve/Symbols.java [hunk 7]
```diff
@@ -125,6 +125,11 @@ public String toString() {
...
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Symbols.java]: MAJOR Reduce this anonymous class number of lines from 21 to at most 20, or make it a named class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1188)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**